src/rule.hpp

```diff
@@ -21,33 +21,116 @@
 
 namespace ddwaf {
 
+/**
+ * Flags that control rule behavior
+ */
+enum class rule_flags : uint8_t {
+    none = 0,
+    generate_event = 1 << 0, // Generate an event when rule matches
+    keep_outcome = 1 << 1    // Prevent the outcome from being dropped on sampling
+};
+
+// Enable bitwise operations on rule_flags
+constexpr ddwaf::rule_flags operator|(ddwaf::rule_flags a, ddwaf::rule_flags b) noexcept
+{
+    return static_cast<ddwaf::rule_flags>(static_cast<uint8_t>(a) | static_cast<uint8_t>(b));
+}
+
+constexpr ddwaf::rule_flags operator&(ddwaf::rule_flags a, ddwaf::rule_flags b) noexcept
+{
+    return static_cast<ddwaf::rule_flags>(static_cast<uint8_t>(a) & static_cast<uint8_t>(b));
+}
+
+constexpr bool operator==(ddwaf::rule_flags a, ddwaf::rule_flags b) noexcept
+{
+    return static_cast<uint8_t>(a) == static_cast<uint8_t>(b);
+}
+
+constexpr bool operator!=(ddwaf::rule_flags a, ddwaf::rule_flags b) noexcept
+{
+    return static_cast<uint8_t>(a) != static_cast<uint8_t>(b);
+}
+
+constexpr bool contains(rule_flags set, rule_flags opt) { return (set & opt) != rule_flags::none; }
+
+enum class rule_verdict : uint8_t { none = 0, monitor = 1, block = 2 };
+enum class rule_source : uint8_t { base = 1, user = 2 };
+
+struct rule_attribute {
+    struct input_target {
+        std::string name;
+        target_index index;
+        std::vector<std::string> key_path;
+    };
+    std::variant<input_target, std::string, uint64_t, int64_t, double, bool> input;
+    std::string output;
```

Comment:
Not a fan of the naming; input/output here are nondescript and are not symmetrical. IIUC one is the attribute value or the source of the attribute value and the other the attribute key.

Reply:
Fixed.